### PR TITLE
Enable greentea tests for us_ticker

### PR DESCRIPTION
**Pull request by fkjagodzinski**
This patch brings a couple of **us_ticker** tests:

- `us_ticker`,
- `us_ticker_lp_ticker_common`, previously named `common_tickers`,
- `us_ticker_lp_ticker_frequency`, previously named `common_tickers_freq`.

I did not enable the **lp_ticker** test; the LP Ticker is out of scope at the moment.
This PR does not enable the higher layer **ticker** tests eventually. This API will be removed in an upcoming PR.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -173,6 +173,10 @@ add_subdirectory(tools/utest)
 
 # HAL tests
 add_subdirectory(tests/mbed_hal/echo EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/ticker/common_tickers EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/ticker/common_tickers_freq EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/ticker/ticker EXCLUDE_FROM_ALL)
```
Comment (MarceloSalazar) [inline CMakeLists.txt:178]: This `ticker/ticker` seems a bit redundant.
Would it be possible to combine it with `common_tickers`?
Alternatively, use another name: e.g. `ticker/xxx`

**Comment by MarceloSalazar** [comment]
Please update the porting guide with details on what each of the tests is verifying and how to run tests.

File: docs/porting/api/us_ticker.md [hunk 2]
```diff
@@ -56,6 +56,18 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
 
 ## Testing
 
-MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker. You can use these tests to validate the correctness of your implementation.
+MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker:
 
-Steps to run the microsecond ticker HAL tests will be provided in the future.
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers` -- verify the defined behavior of the microsecond ticker, see [a paragraph above](#defined-behavior),
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers_freq` -- verify the microsecond ticker is operating at the frequency it specifies (with the use of a host script),
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/us_ticker` -- verify that the microsecond ticker frequency and the microsecond ticker counter width are within correct range, see [a paragraph above](#defined-behavior).
```
Comment (hugueskamba) [inline docs/porting/api/us_ticker.md:63]:
```suggestion
- `MCU-Driver-HAL/tests/mbed_hal/ticker/us_ticker` -- verify that the microsecond ticker frequency and the microsecond ticker counter width are within a correct range, see [paragraph above](#defined-behavior).
```

File: docs/porting/api/us_ticker.md [hunk 3]
```diff
@@ -56,6 +56,18 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
 
 ## Testing
 
-MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker. You can use these tests to validate the correctness of your implementation.
+MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker:
 
-Steps to run the microsecond ticker HAL tests will be provided in the future.
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers` -- verify the defined behavior of the microsecond ticker, see [a paragraph above](#defined-behavior),
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers_freq` -- verify the microsecond ticker is operating at the frequency it specifies (with the use of a host script),
```
Comment (hugueskamba) [inline docs/porting/api/us_ticker.md:62]:
```suggestion
- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers_freq` -- verify that the microsecond ticker is operating at the frequency it specifies (with the use of a host script),
```

File: docs/porting/api/us_ticker.md [hunk 4]
```diff
@@ -56,6 +56,18 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
 
 ## Testing
 
-MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker. You can use these tests to validate the correctness of your implementation.
+MCU-Driver-HAL provides a set of conformance tests for the microsecond ticker:
 
-Steps to run the microsecond ticker HAL tests will be provided in the future.
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers` -- verify the defined behavior of the microsecond ticker, see [a paragraph above](#defined-behavior),
```
Comment (hugueskamba) [inline docs/porting/api/us_ticker.md:61]:
```suggestion
- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers` -- verify the defined behavior of the microsecond ticker, see [paragraph above](#defined-behavior),
```

File: docs/porting/api/us_ticker.md [hunk 5]
```diff
@@ -56,6 +56,18 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
...
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/common_tickers_freq` -- verify the microsecond ticker is operating at the frequency it specifies (with the use of a host script),
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/us_ticker` -- verify that the microsecond ticker frequency and the microsecond ticker counter width are within correct range, see [a paragraph above](#defined-behavior).
+
+MCU-Driver-HAL provides also a set of conformance tests for the generic ticker API that the microsecond ticker inherits:
```
Comment (hugueskamba) [inline docs/porting/api/us_ticker.md:65]:
```suggestion
MCU-Driver-HAL also provides a set of conformance tests for the generic ticker API that the microsecond ticker inherits:
```

File: docs/porting/api/us_ticker.md [hunk 6]
```diff
@@ -56,6 +56,18 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
...
+
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/ticker` -- verify event queue handling and interrupt handling.
+
+You can use these tests to validate the correctness of your implementation.
+
+For a step-by-step guide for building and running the microsecond ticker tests visit the driver implementation repository:
```
Comment (hugueskamba) [inline docs/porting/api/us_ticker.md:71]:
```suggestion
Visit the driver implementation for a step-by-step guide on building and running the microsecond ticker tests:
```

File: tests/mbed_hal/ticker/us_ticker/main.cpp [hunk 7]
```diff
@@ -14,12 +14,13 @@
  * limitations under the License.
  */
 
-#include "mbed.h"
-#include "greentea-client/test_env.h"
-#include "unity.h"
-#include "utest.h"
-#include "us_ticker_api_tests.h"
 #include "hal/us_ticker_api.h"
+#include "us_ticker_api_tests.h"
```
Comment (hugueskamba) [inline tests/mbed_hal/ticker/us_ticker/main.cpp:18]: I am curious as to why all these header files are needed simply to add a reference to `greentea_init_custom_io()`.

Comment (fkjagodzinski) [reply tests/mbed_hal/ticker/us_ticker/main.cpp:18]: This is the commit that adds the custom IO for greentea-client -- https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/6/commits/d77715764f12683f9df75e211b330e26d6f0523b. With only one `#include` directive added.

Other `#include` directives had been updated and sorted in an earlier commit -- https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/6/commits/462a936adc4cf95f7186bdf3324002afc3202085.

**Comment by fkjagodzinski** [comment]
I made a few updates:
- reverted e16875f,
- renamed `common_tickers` to `us_ticker_and_lp_ticker_common`,
- renamed `common_tickers_freq` to `us_ticker_and_lp_ticker_frequency`,
- updated the us_ticker porting guide according to suggestions and added a note that the `ticker` tests do not verify the `us_ticker`.

File: CMakeLists.txt (hunk 1 above)
Comment (fkjagodzinski) [reply CMakeLists.txt]: I renamed the tests in https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/6/commits/871feb703b80686e61e4445d7126a25d450c04f6. I hope the names make more sense now.

File: tests/mbed_hal/ticker/us_ticker/main.cpp (hunk 7 above)
Comment (hugueskamba) [reply tests/mbed_hal/ticker/us_ticker/main.cpp]: Oh I see. Thanks.

File: CMakeLists.txt [hunk 8]
```diff
@@ -173,6 +173,10 @@ add_subdirectory(tools/utest)
 
 # HAL tests
 add_subdirectory(tests/mbed_hal/echo EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/ticker EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/us_ticker EXCLUDE_FROM_ALL)
+add_subdirectory(tests/mbed_hal/us_ticker_and_lp_ticker_common EXCLUDE_FROM_ALL)
```
Comment (hugueskamba) [inline CMakeLists.txt:178]: 🥶I always feel like directories, files and source code identifiers should not have `and` surely we can find a more encompassing name.

Comment (fkjagodzinski) [reply CMakeLists.txt:178]: We could try with `us_ticker_lp_ticker_common`. :smile:

@hugueskamba do you have any alternative in mind? Personally, I don't think `us_ticker_and_lp_ticker_common` is that bad; I'm sure it's better than what we had before (`common_tickers`). It clearly says that both the `us_ticker` and `lp_ticker` can be tested. It also does not mention the upper layer `ticker`.

File: docs/porting/api/us_ticker.md [hunk 9]
```diff
@@ -56,6 +56,14 @@ In addition to the generic `ticker_info_t`, the MCU can also provide compile tim
...
-Steps to run the microsecond ticker HAL tests will be provided in the future.
+- `MCU-Driver-HAL/tests/mbed_hal/ticker/us_ticker` -- verify that the microsecond ticker frequency and the microsecond ticker counter width are within a correct range, see [paragraph above](#defined-behavior),
+- `MCU-Driver-HAL/tests/mbed_hal/us_ticker_and_lp_ticker_common` -- verify the defined behavior of the microsecond ticker, see [paragraph above](#defined-behavior),
+- `MCU-Driver-HAL/tests/mbed_hal/us_ticker_and_lp_ticker_frequency` -- verify that the microsecond ticker is operating at the frequency it specifies (with the use of a host script).
+
+You can use these tests to validate the correctness of your implementation.
+
+Visit the driver implementation for a step-by-step guide on building and running the microsecond ticker tests.
+
+MCU-Driver-HAL also provides a set of conformance tests for the higher layer `ticker` API that utilises either the microsecond ticker or the low power ticker. The `ticker` API adds more features like suspend/resume or event queue handling. The `MCU-Driver-HAL/tests/mbed_hal/ticker` tests do not exercise the microsecond ticker implementation.
```
Comment (MarceloSalazar) [inline docs/porting/api/us_ticker.md:69]: `ticker` is considered a middleware and doesn't make use of hardware timers,  hence we propose to remove it from the HAL API.

File: tests/mbed_hal/ticker/CMakeLists.txt [hunk 10]
```diff
@@ -13,13 +13,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
+set(TEST_TARGET_LIB sdfx-test-ticker)
```
[the remark this answers is not in the document]
Comment (fkjagodzinski) [reply tests/mbed_hal/ticker/CMakeLists.txt:16]: I think this makes a bit more sense if you look at [the CMake file in the implementation repo](https://github.com/MCU-Driver-HAL/MCU-Driver-ST/pull/3/files#diff-3ea694748c03b05fd56282988f0d122983d07e31ce4700d5bd88b95975c46df8R21-R39) or [a target with a longer name](https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/6/files#diff-c260750f6de41f6c30609e39c20a6b5f0ead957a6c21985720b61900c822660eR16).

Do you consider this bad practice?

**Comment by fkjagodzinski** [comment]
New updates:
- removed **ticker** tests,
- shortened test names
    - `us_ticker_lp_ticker_common`,
    - `us_ticker_lp_ticker_frequency`,
- removed the note about the `ticker` tests from the `us_ticker` porting guide,
- addressed review comments
    - removed updates to `echo` test,
    - simplified CMake files,
    - moved a call to `greentea_init_custom_io()` from test setup to `main()`.

File: CMakeLists.txt (hunk 8 above)
Comment (fkjagodzinski) [reply CMakeLists.txt]: While I wasn't completely serious about it at first, now I find `us_ticker_lp_ticker_common` to make sense. I propose to go with this name.

File: tests/mbed_hal/echo/main.cpp [hunk 11]
```diff
@@ -1,19 +1,19 @@
-/*
- * Copyright (c) 2013-2016, ARM Limited, All Rights Reserved
+/* Copyright (c) 2013-2021 Arm Limited
  * SPDX-License-Identifier: Apache-2.0
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may
- * not use this file except in compliance with the License.
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (fkjagodzinski) [reply tests/mbed_hal/echo/main.cpp]: Done; updates to the echo test are no longer present in this PR.

File: tests/mbed_hal/ticker/CMakeLists.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (fkjagodzinski) [reply tests/mbed_hal/ticker/CMakeLists.txt]: Done; CMake files simplified.

File: tests/mbed_hal/ticker/main.cpp [hunk 12]
```diff
@@ -2452,6 +2451,7 @@ static const case_t cases[] = {
 
 static utest::v1::status_t greentea_test_setup(const size_t number_of_cases)
 {
+    greentea_init_custom_io();
```
[the remark this answers is not in the document]
Comment (fkjagodzinski) [reply tests/mbed_hal/ticker/main.cpp]: Done.

File: docs/porting/api/us_ticker.md (hunk 9 above)
Comment (fkjagodzinski) [reply docs/porting/api/us_ticker.md]: OK, will be removed in an upcoming PR. **ticker** tests are no longer enabled by this PR.

**Review by MarceloSalazar — APPROVED** [review]

**Review by ithinuel — APPROVED** [review]

**Pull request merged**